client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -525,6 +525,116 @@ private RetryAndThrowConditionally createRetryConfig(final ControllerImplConfig
+                    log.warn(requestId, "Stream does not exist: {}", streamName);
+                    throw new IllegalArgumentException("Stream does not exist: " + streamName);
+                case SUBSCRIBER_EXISTS:
+                    log.warn(requestId, "Subscriber {} for stream {}/{} already exists {}.", subscriber, scope, streamName);
+                    throw new IllegalArgumentException("Stream does not exist: " + streamName);
+                case SUCCESS:
+                    log.info(requestId, "Successfully updated stream: {}", streamName);
+                    return true;
+                case UNRECOGNIZED:
+                default:
+                    throw new ControllerFailureException("Unknown return status adding subscriber " + subscriber
+                            + " " + x.getStatus());
+            }
+        }).whenComplete((x, e) -> {
+            if (e != null) {
+                log.warn(requestId, "addSubscriber {} for stream {}/{} failed: ", subscriber, scope, streamName, e);
+            }
+            LoggerHelpers.traceLeave(log, "addSubscriber", traceId, subscriber, requestId);
+        });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> removeSubscriber(String scope, String streamName, String subscriber) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkNotNull(subscriber, "subscriberId");
+        final long requestId = requestIdGenerator.get();
+        long traceId = LoggerHelpers.traceEnter(log, "removeSubscriber", subscriber, requestId);
+
+        final CompletableFuture<RemoveSubscriberStatus> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<RemoveSubscriberStatus> callback = new RPCAsyncCallback<>(requestId, "removeSubscriber", scope, streamName, subscriber);
+            new ControllerClientTagger(client, timeoutMillis).withTag(requestId, "removeSubscriber", scope, streamName)
+                    .removeSubscriber(ModelHelper.decode(scope, streamName, subscriber), callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(x -> {
+            switch (x.getStatus()) {
+                case FAILURE:
+                    log.warn(requestId, "Failed to update stream: {}", streamName);
+                    throw new ControllerFailureException("Failed to add subscriber: " + subscriber);
+                case SCOPE_NOT_FOUND:
+                    log.warn(requestId, "Scope not found: {}", scope);
+                    throw new IllegalArgumentException("Scope does not exist: " + scope);
+                case STREAM_NOT_FOUND:
+                    log.warn(requestId, "Stream does not exist: {}", streamName);
+                    throw new IllegalArgumentException("Stream does not exist: " + streamName);
+                case SUBSCRIBER_NOT_FOUND:
+                    log.warn(requestId, "Subscriber does not exist: {} for stream {}/{}", subscriber, scope, streamName);
```

Comment:
same as above

Reply:
remove is different from above case. since we are not doing conditional remove, but is predicated on the subscriber's name. it should perhaps not throw the exception. 

in fact i would say that this should not even be a response code for remove api from idempotent case's perspective.

Reply:
All other delete APIs return response codes. I think it is important to keep APIs consistent for better user experience...

Reply:
Ok changed return status for SUBSCRIBER_NOT_FOUND to false.